bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,220 @@
+
+            repo.fetchAll();
+            var depsFile = repo.root().resolve(depsFileName);
+
+            var orderedBranches = orderedBranches(repo, depsFile);
+            log.info("Merge order " + orderedBranches);
+            for (var branch : orderedBranches) {
+                log.info("Processing branch " + branch + "...");
+                repo.checkout(branch);
+                var parents = dependencies(depsFile).collect(Collectors.toSet());
+                List<String> failedMerges = new ArrayList<>();
+                boolean progress;
+                boolean failed;
+                do {
+                    // We need to attempt merge parents in any order that works. Keep merging
+                    // and pushing, until no further progress can be made.
+                    progress = false;
+                    failed = false;
+                    for (var parentsIt = parents.iterator(); parentsIt.hasNext();) {
+                        var parent = parentsIt.next();
+                        try {
+                            mergeIfAhead(repo, branch, parent);
+                            progress = true;
+                            parentsIt.remove(); // avoid doing pointless merges
+                        } catch(IOException e) {
+                            log.severe("Merge with " + parent + " failed. Reverting...");
+                            repo.abortMerge();
+                            failedMerges.add(branch + " <- " + parent);
+                            failed = true;
+                        }
+                    }
+                } while(progress && failed);
+
+                if (!failedMerges.isEmpty()) {
+                    throw new IOException("There were failed merges:\n" + failedMerges);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        log.info("Ending topobot run");
+    }
+
+    private static Stream<Branch> dependencies(Path depsFile) throws IOException {
+        if (Files.exists(depsFile)) {
+            var lines = Files.readAllLines(depsFile).stream().filter(s -> !s.isEmpty()).collect(Collectors.toList());
+            if (lines.size() > 1) {
+                throw new IllegalStateException("Multiple non-empty lines in " + depsFile.toString() + ": "
+                        + String.join("\n", lines));
+            }
+            return Stream.of(lines.get(0).split(" ")).map(Branch::new);
+        } else {
+            return Stream.of(new Branch("master"));
+        }
+    }
+
+    private List<Branch> orderedBranches(Repository repo, Path depsFile) throws IOException {
+        List<Edge> deps = new ArrayList<>();
+        for (var branch : branches) {
+            repo.checkout(branch);
+            dependencies(depsFile).forEach(dep -> deps.add(new Edge(dep, branch)));
+        }
+        return TopologicalSort.tsort(deps).stream()
+            .filter(branch -> !branch.name().equals("master"))
```

Comment:
If you want you can use `.filter(branch -> !branch.equals(repo.defaultBranch()))`